### PR TITLE
Invalid launch options no longer cause Xenia to exit silently

### PR DESCRIPTION
**Pull request by Significantbits**
This pull request aims to fix issue #1445  .  The cause of this issue is that the PrintHelpAndExit function was only writing to stdout out instead of notifying the user via a Message Box or something similar. 

This function gets called under two circumstances.

1.) The user gave the --help argument to Xenia. 
2.) The user gave invalid launch options to Xenia.

In order to differentiate between these cases, a boolean was added as an input to the function. This allowed me to put in a different message box for each of these cases. The --help message box displays the usage information, and the invalid argument message box tells the user that they gave invalid launch options then displays the usage information. 

image

image

This is my first pull request to this project, so hopefully everything follows the contribution guidelines. I tried to be careful, but please let me know if I did something wrong.

### CODE REVIEW

**Comment by jgoyvaerts** [comment]
Have you read this issue https://github.com/xenia-project/xenia/issues/1516 and tested your solution in mingw/bash based terminals? Do we really want both stdout AND a message box?

**Comment by jgoyvaerts** [comment]
Also, take a look at logging.cc, there's already a message box being shown there in case of a fatal error. Can you reuse that function or write a message box function there instead of jamming the messagebox code in the middle of this cvar file?

**Comment by Significantbits** [comment]
Please ignore the last push. I messed up. Fixing it now.

File: src/xenia/base/main_win.cc [hunk 1]
```diff
@@ -58,6 +58,16 @@ void AttachConsole() {
   fp = _fdopen(con_handle, "w");
   *stderr = *fp;
   setvbuf(stderr, nullptr, _IONBF, 0);
+
+  char* buf = nullptr;
```
Comment (jgoyvaerts) [inline src/xenia/base/main_win.cc:62]: Why these changes?

File: src/xenia/kernel/util/shim_utils.h [hunk 2]
```diff
@@ -252,7 +252,7 @@ class PrimitivePointerParam : public ParamBase<uint32_t> {
   }
   T value() const { return *host_ptr_; }
   operator T() const = delete;
-  operator xe::be<T>*() const { return host_ptr_; }
+  operator xe::be<T> *() const { return host_ptr_; }
```
Comment (jgoyvaerts) [inline src/xenia/kernel/util/shim_utils.h:255]: Any reason for this change?

File: src/xenia/base/cvar.cc [hunk 3]
```diff
@@ -60,7 +79,11 @@ void ParseLaunchArguments(int& argc, char**& argv,
 
     auto result = options.parse(argc, argv);
     if (result.count("help")) {
-      PrintHelpAndExit();
+      if (xe::has_console_attached()) {
+        PrintHelpAndExit();
+      } else {
+        ExitWithNotification(true);
```
Comment (jgoyvaerts) [inline src/xenia/base/cvar.cc:85]: Instead of the ExitWithNotification function, make 1 or 2 functions in logging.cc (or somewhere else if there's a better place): ShowInfoMessageBox(string) and ShowErrorMessageBox(string), and use those directly

File: src/xenia/kernel/util/shim_utils.h (hunk 2 above)
Comment (Significantbits) [reply src/xenia/kernel/util/shim_utils.h:255]: I think that happened when I did xb format. I didn't change that myself. I'll push again without these changes.

File: src/xenia/base/main_win.cc (hunk 1 above)
Comment (Significantbits) [reply src/xenia/base/main_win.cc:62]: This is to make it work with a bash environment or Windows terminal environment. I looked at issue #1516 like you said to. I tried implement the solution to redirect stdout and stderr to the parent process, but this broke the bash terminal environment. So, I check if the SHELL environment variable exists. If it doesn't exists, I assume a Windows terminal environment.

MSYS2
image
image

Windows Terminal
image

Comment (Significantbits) [reply src/xenia/base/main_win.cc:62]: However, if there is no console attached then you will get the message box pops up. For example, if you give invalid arguments in Visual Studio, then you will see the Message Box.

Comment (jgoyvaerts) [reply src/xenia/base/main_win.cc:62]: What do you mean to make it work? You check with xe::has_console_attached if there's a console attached, so what does this specific piece of code in main_win.cc do? What happens or doesn't work if you remove it?

Comment (Significantbits) [reply src/xenia/base/main_win.cc:62]: xe::has_console_attched proves true if you launch from a Windows Terminal or bash environment. However, if you don't redirect stdout and stderr you won't see the print message in the Windows terminal, and Xenia will just exit silently like before. If you redirect stdout and stderr, then you won't see the message if Xenia is launched from a bash environment.  This code checks to see which environment Xenia was launched from and either redirects stdout and stderr, or doesn't redirect it.

So, in short, if you remove it, Xenia will still exit silently if launched from a Windows Terminal. 

I thought just getting a print message in the console would be better than a Message Box if xe::has_console_attached is true.

Comment (jgoyvaerts) [reply src/xenia/base/main_win.cc:62]: If I can see your screenshot correctly, you're writing to a terminal which is not expecting output, which messes up the console for no reason.

As I said in #1516 , the windows Cmd and Powershell do not wait for output when launching Xenia through the command line. Instead of finding a quirky workaround to still output to a console which isn't expecting output, isn't it easier to fix has_console_attached to return false in those cases? (See AttachConsole in main_win.cc and feel free to improve it)

Comment (Significantbits) [reply src/xenia/base/main_win.cc:62]: Okay. I'll fix has_console_attached so that a Message Box will pop up if you are using the a Windows terminal.

File: src/xenia/base/cvar.cc (hunk 3 above)
Comment (Significantbits) [reply src/xenia/base/cvar.cc]: I did this in the latest push. Sorry for misunderstanding the previous comment on this.

File: src/xenia/base/cvar.cc [hunk 4]
```diff
@@ -24,6 +27,20 @@ cxxopts::Options options("xenia", "Xbox 360 Emulator");
 std::map<std::string, ICommandVar*>* CmdVars;
 std::map<std::string, IConfigVar*>* ConfigVars;
 
+/*void ExitWithNotification(bool help) {
```
Comment (jgoyvaerts) [inline src/xenia/base/cvar.cc:30]: There's commented code here, you should probably clean this up

**Comment by jgoyvaerts** [comment]
If I was you I'd split up the commit in 2 or 3 commits
- One for the fixing/improving of has_console_attached
- One for the adding of the helper methods in logging.cc
- One for the changing of the behavior in cvar.cc

File: src/xenia/base/cvar.cc (hunk 4 above)
Comment (Significantbits) [reply src/xenia/base/cvar.cc]: Sorry about that. Should be gone now.
